src/utils/GISCache.ts

```diff
@@ -0,0 +1,424 @@
+
+const maxAge = 86400 * 7 // seconds
+
+class GISCache {
+  private readonly db: Database
+
+  constructor () {
+    this.db = sqlite3('cache.db') 
+
+    this.initTables()
+    this.update()
+  }
+
+  get context (): Database {
+    return this.db
+  }
+
+  async update () {
+    const record = this.db.prepare("SELECT value FROM params WHERE key = 'timestamp'").get()
+    const now = Math.round(+new Date() / 1000)
+
+    try {
+      if (!record || !record.value) {
+        console.log('creating cache...')
+        this.db.exec('INSERT INTO params ("key", value)VALUES(\'timestamp\', -1);')
+        this.repopulate(await this.getUpdatedData())
+        this.updateTimestamp()
+      } else if ( +record.value + maxAge < now) {
+          // const list = this.getUpdatedData()
+          // todo : check which data can be update
+          console.log('updating cache...')
+          this.updateTimestamp()
+      } else {
+        console.log('cache is up to date')
+      }
+
+    } catch (e) {
+      console.error('CacheUpdateError', e)
+    }
+
+  }
+
+  updateTimestamp () {
+    const now = Math.round(+new Date() / 1000)
+    this.db
+    .prepare('UPDATE params SET value = ? WHERE "key" = \'timestamp\' LIMIT 1')
+    .run(now)
+  }
+
+  async getUpdatedData (): Promise<BuildingMulti[]> {
+    return (await GISService.getAll()).features.map(f => {
+      const props = f.properties
+      const styleFR = props.STYLE_FR ? props.STYLE_FR.split(/\s*,\s*/) : null // props.STYLE_FR?.split( ts-3.7
+      const styleNL = props.STYLE_NL ? props.STYLE_NL.split(/\s*,\s*/) : null
+      const intervenants = props.INTERVENANTS ? props.INTERVENANTS.split(/(\s*),\s*/).map(inter => {
+        const data = inter.trim().match(/^(?<name>[^\d]+)(?:\((?<start>\d{4})(?:-(?<end>\d{4}))?\))?$/)
+        return (data && data.groups ? {
+          name: data.groups ? data.groups.name.trim() : inter,
+          startYear: data.groups.start || null,
+          endYear: data.groups.end || null
+        } : { name: inter })as IntervenantAPI
+      }).filter(inter => inter.name.replace(/\s+/g, '') !== '') : null
+      return {
+        zipCode: props.CITY,
+        cityFR: props.CITIES_FR,
+        cityNL: props.CITIES_NL,
+        id: props.ID,
+        idBatiCMS: props.ID_BATI_CMS,
+        image: props.FIRSTIMAGE,
+        intervenants,
+        nameFR: props.NOM_FR,
+        nameNL: props.NOM_NL,
+        number: props.NUMBER,
+        streetFR: props.STREET_FR,
+        streetNL: props.STREET_NL,
+        styleFR,
+        styleNL,
+        typoID: props.TYPO,
+        typoFR: props.TYPO_FR,
+        typoNL: props.TYPO_NL,
+        urlFR: props.URL_FR,
+        urlNL: props.URL_NL,
+        gpsLon: f.geometry.coordinates[0],
+        gpsLat: f.geometry.coordinates[1]
+      } as BuildingMulti
+    })
+  }
+
+  repopulate (data: BuildingMulti[]) {
+    this.truncateTables(true)
+    const cities = {} as { [key: string]: City }
```

Comment:
I'm a bit confused about `{ [key: string]: City }`.
That might be correct but isn't it more clear/readable to use `Record<string, City>` ?